### ISSUE TITLE
Importing API types pulls in lots of dependencies

### ISSUE THREAD

**Issue by JoelSpeed** [kubernetes-sigs/cluster-api#9011]
### What steps did you take and what happened?

Create a simple program, as below and run `go mod tidy`:
```
package main

import (
	"fmt"

	capiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
)

func main() {
	myCluster := &capiv1beta1.Cluster{}

	fmt.Printf("My Cluster: %+v\n", myCluster)
}
```
This results in a [go.mod](https://github.com/JoelSpeed/capi-api-import-test/blob/df4df473e547141d354718f6c3e1cd2866f6b7b2/go.mod) with 56 dependencies. Included in this list is `controller-runtime` which, is also imported by a lot of consumers of the API types. `controller-runtime` has evolved a lot over time and, because it is being imported by the APIs currently, means that any project consuming the API types must also import the same controller-runtime as the rest of CAPI.

There are use cases where, building a consumer, I do not need to use any of the code (defaulting, validation etc) from CAPI and as such, do not need such a restriction on my `controller-runtime` version.

If we modify the way the `SchemeBuilder` works and move the [webhooks to a separate package](https://github.com/kubernetes-sigs/cluster-api/compare/main...JoelSpeed:cluster-api:remove-deps-from-api?expand=1), then we can reduce the required imports for the API package down to just [31 dependencies](https://github.com/JoelSpeed/capi-api-import-test/blob/bca5c2abf8d9397e66b8e45d95a755ff7f649347/go.mod#L1), and, these dependencies do not include `controller-runtime` or other fast moving/changing packages. There are still obviously a good number of deps in there from K8s paths, but these are mostly API machinery related and have pretty stable APIs.

I think if we can separate the webhooks out in this way (I notice Cluster and ClusterClass already are?) then I think this would make consuming these API types in other projects a lot easier.

Note how `k8s.io/api` does not include much code with the API types for this reason as well.

The alternative solution for consumers is to copy/paste the types out to their own repos which is less ideal.

### What did you expect to happen?

API types should not be tied to many dependencies, especially not controller-runtime.

### Cluster API version

main branch

### Kubernetes version

N/A

### Label(s) to be applied

/kind bug
One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels.

**Comment by vincepri** [comment]
This all sounds good, a few additional comments would be:

- We should have a clear way on how to structure webhooks in our API packages across Cluster API and its ecosystem of providers
- Separate the SchemeBuilder changes from the webhook ones
- Open an issue in Kubebuilder to potentially change the default way new projects or packages are generated

Thoughts?
cc @CecileRobertMichon @fabriziopandini @sbueringer

**Comment by JoelSpeed** [comment]
> We should have a clear way on how to structure webhooks in our API packages across Cluster API and its ecosystem of providers

Agreed, I don't have context on why `Cluster` and `ClusterClass` are separate, but, their pattern works well. Do you think we need to have versioned defaulting/validation? If so, does it make sense to have the webhooks in versioned packages as in my POC branch?

If they don't need to be versioned, having them centrally as for `Cluster` and `ClusterClass` would work.

> Separate the SchemeBuilder changes from the webhook ones

Ack yes, I can make a PR for this separately so it's considered as a different part of the same effort

---

I think it would be good to have a consistent approach across CAPI and the related repos, if there is a pattern we can come up with that we can document then I would like to also pursue updating the provider APIs to mitigate this issue there as well

**Comment by sbueringer** [comment]
> Agreed, I don't have context on why Cluster and ClusterClass are separate, but, their pattern works well. Do you think we need to have versioned defaulting/validation? If so, does it make sense to have the webhooks in versioned packages as in my POC branch?

The reason why Cluster and ClusterClass are separate is the following:
* we couldn't just implement the Validate / Default methods on the types directly because we needed a client in those webhooks
* So we already had to use CustomDefaulter / CustomValidator
* I didn't want to add even more dependecies to our API package so I suggested to move them to a separate package

I agree that we can just follow the same pattern for all webhooks.

> If they don't need to be versioned, having them centrally as for Cluster and ClusterClass would work.

The validate / default funcs / webhooks don't have to be versioned. Whenever we introduced a new apiVersion we moved the methods from the previous apiVersion to the new one :).


So overall sounds good to me. I'm absolutely in favor!

**Comment by CecileRobertMichon** [comment]
+1 

Agree anything we can do to limit importing `controller-runtime` only where necessary is an improvement

**Comment by sbueringer** [comment]
> We discussed this issue at the backlog grooming on Oct 29th 2021, the consensus was that for now we're not able to fully decouple our API types from the controller runtime dependencies due to conversion webhooks which are strictly tied to our types.

Just to double-check. I think I might have brought this up at the time. If I now take a look at our current code this doesn't seem to be an issue as as we only depend on ""sigs.k8s.io/controller-runtime/pkg/conversion" (for conversion.Hub) in old apiVersions. So it won't be a problem for cleaning up dependencies of the latest apiVersion.

**Comment by fabriziopandini** [comment]
+1 from my side, thanks for reviving this discussion @JoelSpeed!

**Comment by JoelSpeed** [comment]
I did a bit of work on the code for this today and came across a small issue.

Currently the package `internal/test/builder` imports the defaulting functions from `clusterv1`, if we move these to `webhooks`, it creates an import cycle because the tests are importing the `internal/test/builder` to run tests on the webhooks.

I can see 3 possible paths forward:
* Keep the `Default()` functions inside the `clusterv1` package and require that they do not use any import from controller-runtime (not ideal, don't know that this is even possible)
* Remove the defaulting bit from the builder (it's for MHC and MD) and handle the defaulting in the tests where those are needed (there are 9 instances for MHC, 1 for MD)
*  Try to move the webhook tests to a separate go package, `webhooks_test`, which may solve the issue as well (loses access to private functions)

Any preferred method from those or alternative suggestions from anyone?

**Comment by killianmuldoon** [comment]
>Remove the defaulting bit from the builder (it's for MHC and MD) and handle the defaulting in the tests where those are needed (there are 9 instances for MHC, 1 for MD)

From a quick look it should be relatively easy to move the defaulting to the tests themselves rather than the builders. There's only a few instances of this usage. @chrischdi - I think you added this when moving these tests to envtest, is there something I'm not thinking of that could stand in the way?

**Comment by JoelSpeed** [comment]
> Remove the defaulting bit from the builder (it's for MHC and MD) and handle the defaulting in the tests where those are needed (there are 9 instances for MHC, 1 for MD)

Going to play with this idea then, will report back if I hit more issues. Expect some PRs over the coming days

**Comment by sbueringer** [comment]
If I got it right we already had and solved this issue before. We have a separate test package here for those kind of tests: https://github.com/kubernetes-sigs/cluster-api/tree/3e16abc3df71314bbaee770846beba45c0eb170d/internal/webhooks/test

**Comment by chrischdi** [comment]
> > Remove the defaulting bit from the builder (it's for MHC and MD) and handle the defaulting in the tests where those are needed (there are 9 instances for MHC, 1 for MD)
> 
> From a quick look it should be relatively easy to move the defaulting to the tests themselves rather than the builders. There's only a few instances of this usage. @chrischdi - I think you added this when moving these tests to envtest, is there something I'm not thinking of that could stand in the way?

Nope, looks good. At the time of implementing it was just the most straightforward one. Should be totally fine to default at the test and remove that from the builder 👍 

@sbueringer because we just chatted about the `depguard` linter: maybe that one would be useful if the config is flexible enough, to prevent regressions if this issue got solved 🙂

**Comment by JoelSpeed** [comment]
I've just posted two PRs to show the progress I've made so far on the main Cluster v1beta1 APIs, the first of which for scheme builder should be ready for review, the second I think is ready for review apart from I know I have to sort the MachineDeployment defaulting in tests, otherwise I think that should be fairly good to go

**Comment by JoelSpeed** [comment]
Both PRs are now ready for review and, combined, clear the deps for the main APIs (Machine, MS, MD, Cluster etc). Will start to look at the exp APIs next

**Comment by JoelSpeed** [comment]
Have added a TODO list with my initial thoughts on scope here, any additional points we need for this issue?

**Comment by sbueringer** [comment]
Seems ~ good to me. Does your list cover all those APIs?

stable
./api
./controlplane/kubeadm/api
./bootstrap/kubeadm/api

exp
./exp/api
./exp/addons/api
./exp/ipam/api
./exp/runtime/api

clusterctl
./cmd/clusterctl/api (only one usage of scheme.Builder)

test
./test/infrastructure/docker/api
./test/infrastructure/docker/exp/api
./test/infrastructure/inmemory/api

Totally fine to focus on stable + exp + maybe clusterctl (because it's a quickwin) and then create a follow-up issue for test (I would mostly migrate test as well for consistency across the entire code base)

P.S. I'll try to find some time to review your PRs soon. Pretty low bandwidth at the moment

**Comment by JoelSpeed** [comment]
I've updated the list to be very explicit, will keep track of the PRs in the list as we go along as well

**Comment by sbueringer** [comment]
@killianmuldoon Do you have some time to replicate this issue for CAPV and work on it going forward? (implementation is not urgent, doesn't have it into CAPV v1.8)

**Comment by killianmuldoon** [comment]
>@killianmuldoon Do you have some time to replicate this issue for CAPV and work on it going forward? (implementation is not urgent, won't make it / doesn't have to make it into CAPV v1.8)

By that you mean removing the controller-runtime package from the CAPV API packages?

**Comment by sbueringer** [comment]
> By that you mean removing the controller-runtime package from the CAPV API packages?

Basically implementing the same in CAPV that Joel is implementing in core CAPI. So yes, if I got your question correctly :)
(although what Joel is implementing only removes the dependency from the v1beta1 packages, but that's enough)

**Comment by killianmuldoon** [comment]
Sure thing - I'll create an issue and assign it to myself.

**Comment by alexander-demicev** [comment]
We are planning to do the same with the CAPI operator. Webhooks are already in a separate package. We have a dependency on controller runtime in the API itself that will be removed in the new API version. This change will definitely simplify consuming CAPI/CAPI operator programmatically.

**Comment by sbueringer** [comment]
Just found this tool https://pkg.go.dev/k8s.io/code-generator/cmd/import-boss

At a first glance this could help us to guard against regressions

**Comment by fabriziopandini** [comment]
+1 to add import-boss as a final step of this effort

**Comment by JoelSpeed** [comment]
Added a new PR to move the schemebuilder over for the experimental APIs, https://github.com/kubernetes-sigs/cluster-api/pull/9185

**Comment by JoelSpeed** [comment]
Another PR to remove the final reliance on scheme builder stuff, https://github.com/kubernetes-sigs/cluster-api/pull/9266

**Comment by sbueringer** [comment]
Talked to Ankita, she'll take over `./bootstrap/kubeadm/api`

**Comment by odvarkadaniel** [comment]
> Talked to @Ankitasw, she'll take over `./bootstrap/kubeadm/api`. Thank you!

I have got that in the PR already. If needed, I will just remove the commit.

**Comment by Ankitasw** [comment]
No worries, I can pick .exp/api if not yet picked.

**Comment by sbueringer** [comment]
Sounds good. Sorry didn't see it

**Comment by Ankitasw** [comment]
Will pick `.exp/addons/api` next

**Comment by sbueringer** [comment]
Let's do a final round at the end to add import restrictions everywhere. I forgot to ask for them at least a few times and we don't have them in the packages we added earlier. Should be good enough to just add them once we moved everything.

**Comment by Ankitasw** [comment]
Will pick `./test/infrastructure/docker/api` now

**Comment by Ankitasw** [comment]
Also picking `./test/infrastructure/docker/exp/api` and `./test/infrastructure/inmemory/api`

**Comment by sbueringer** [comment]
Thank you very much!

**Comment by Ankitasw** [comment]
@sbueringer @JoelSpeed I think the webhook in`./test/infrastructure/docker/exp/api` is a dummy one, and doesn't do anything or used anywhere, so it might not be required to move the packages. What do you folks suggest?

**Comment by sbueringer** [comment]
Nice work everyone!!

As far as I can tell we should be done - apart from documentation. 

@JoelSpeed Would be good if you can maybe do a final verification if the transitive dependencies look good from your side (in case you didn't do it already)

**Comment by JoelSpeed** [comment]
> @JoelSpeed Would be good if you can maybe do a final verification if the transitive dependencies look good from your side (in case you didn't do it already)

Based on https://github.com/JoelSpeed/capi-api-import-test/commit/8800fd3945ddaeedca26c83015f166f897569468 we are doing pretty well across the board, however, the Addons API, which I added separately, https://github.com/JoelSpeed/capi-api-import-test/commit/8c6aa3d6bfcfde755a9250eba226f09a814a0f75, adds a bunch of deps including controller runtime back in.

I see the import restrictions are in place for the addons, I guess it doesn't work transitively?

**Comment by JoelSpeed** [comment]
Seems we are importing `sigs.k8s.io/cluster-api/util` [here](https://github.com/kubernetes-sigs/cluster-api/blob/b3b9928c06a1a3cc1b3aea9c419ca37492536317/exp/addons/api/v1beta1/clusterresourcesetbinding_types.go#L109-L113) which is bringing in the additional imports, even though the code it uses within that module doesn't itself use CR

**Comment by killianmuldoon** [comment]
Yeah - looks like we need a refactor of some of that code - I'll have a quick go and see if it's straightforward.

**Comment by sbueringer** [comment]
Added a new sub task:
```
[ ] Add verification to avoid regressions (e.g. https://github.com/kubernetes-sigs/cluster-api/pull/9482#issuecomment-1730060650)
```

**Comment by fabriziopandini** [comment]
@sbueringer @enxebre should we close this issue? the development work is completed and we have our code as a documentation/example for how to do the same in providers (if they want to)

**Comment by sbueringer** [comment]
I think we have to implement " Add verification to avoid regressions" otherwise we have 0 protection against transitive dependencies sneaking in. Which can happen very easily.

One idea how this can be done was suggested here: https://github.com/kubernetes-sigs/cluster-api/pull/9482#issuecomment-1730060650 (same PR shows how easy it is to get transitive dependencies)

**Comment by stmcginnis** [comment]
It looks like the majority of the TODOs have been completed. Can this issue be closed now?

**Comment by sbueringer** [comment]
https://github.com/kubernetes-sigs/cluster-api/issues/9011#issuecomment-1914532740

I have something coming up (https://github.com/kubernetes-sigs/controller-runtime/pull/3335) that will probably allow us to move CR conversion code out of the API package and then introduce a separate API go module with minimal dependencies. That will make it easy to track which dependencies our APIs actually have

**Comment by sbueringer** [comment]
To give a rough idea on how I'm planning to close out this issue:
* [x] [A long time ago] Move defaulting and validation webhooks out of api/
* [x] https://github.com/kubernetes-sigs/controller-runtime/pull/3335
  * Allows us to implement conversion outside of API packages (ConvertTo/ConvertFrom/Hub) once we pickup CR v0.23 (early in the CAPI v1.13 cycle)
* [ ] Remove v1alpha3/v1alpha4: https://github.com/kubernetes-sigs/cluster-api/issues/11919
* [ ] Move index package out of api/
* [ ] Move conversion & conversion tests out of api/
  * Includes creating a new conversion fuzz test util
* [ ] Move sigs.k8s.io/cluster-api/errors errors into api/
  * This package is only used for v1beta1 and will be removed with v1beta1
* [ ] Introduce a new sigs.k8s.io/cluster-api/api go module
* [ ] [Note to myself] Same for CAPV
* [ ] [TBD but probably not] Try to move zz_generated.conversion.go out of api/

cc @fabriziopandini

**Comment by sbueringer** [comment]
To give a rough idea on how I'm planning to close out this issue:
* [x] [A long time ago] Move defaulting and validation webhooks out of api/
* [x] https://github.com/kubernetes-sigs/controller-runtime/pull/3335
  * Allows us to implement conversion outside of API packages (ConvertTo/ConvertFrom/Hub) once we pickup CR v0.23 (early in the CAPI v1.13 cycle)
* [x] Remove v1alpha3/v1alpha4: https://github.com/kubernetes-sigs/cluster-api/issues/11919
* [x] Introduce api module
  * [x] Remove dependencies between "api" and "main" module
  * [x] Move conversion & conversion tests out of api/
    * [x] Block all CR imports
    * [x] Also adjust & improve convert/resource.go: https://github.com/kubernetes-sigs/cluster-api/pull/12843
  * [x] Introduce a new sigs.k8s.io/cluster-api/api go module
  * [x] Add verification to avoid regressions (https://github.com/kubernetes-sigs/cluster-api/issues/9011)
  * [x] Adjust release workflow
  * PR: https://github.com/kubernetes-sigs/cluster-api/pull/13762

Follow-up:
* [-] [TBD but probably not] Try to move zz_generated.conversion.go out of api/ => won't do for now until there is strong demand

cc @fabriziopandini @JoelSpeed @chrischdi (I'll bring it up in office hours today)

PR: https://github.com/kubernetes-sigs/cluster-api/pull/13762

**Comment by fabriziopandini** [comment]
Great to see a path forward for this issue, a big Kudos to @sbueringer and the controller runtime team to make this possible

**Comment by sbueringer** [comment]
Now that we introduced a separate api module I think we can call this done.

If someone wants to follow-up for documentation feel free.

/close
